### ISSUE TITLE
Slow start up time as compared to `alacritty`: ideas for fixes

### ISSUE THREAD

**Issue by ctrlcctrlv** [kovidgoyal/kitty#4292]
<sub>(I've had an interesting few days, given a post I never intended to receive wide circulation (it's not even a blog post, just [a GitHub Gist I passed around](https://gist.github.com/ctrlcctrlv/978b3ee4f55d4b4ec415a985e01cb1c9) a few years ago) was posted by someone I don't know to Hacker News, where it trended, and has attracted comments by, among others, @primalmotion. This bug report originates from there.)</sub>

Now, kitty is well known for being the fastest terminal emulator by many metrics. However, an area where it apparently lags behind is startup time. I believe we can cut the startup time pretty easily.

Sometimes it starts up rather quickly for me compared to alacritty:

```
[fred@🍇葡萄🍇~]$ time env -u WAYLAND_DISPLAY alacritty -e echo

real	0m0.165s
user	0m0.143s
sys	0m0.049s
[fred@🍇葡萄🍇~]$ time kitty --instance-group debug9 echo
[332 01:13:15.006344] Ignoring invalid config line: url_excluded_characters

real	0m0.251s
user	0m0.180s
sys	0m0.078s
```

Other times it takes longer, and when this happens I narrowed it to `load_all_shaders` being the slowest thing to do:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/bin/../lib/kitty/__main__.py", line 153, in <module>
    main()
  File "/usr/bin/../lib/kitty/__main__.py", line 147, in main
    kitty_main()
  File "/usr/bin/../lib/kitty/kitty/main.py", line 349, in main
    _main()
  File "/usr/bin/../lib/kitty/kitty/main.py", line 342, in _main
    run_app(opts, cli_opts, bad_lines)
  File "/usr/bin/../lib/kitty/kitty/main.py", line 183, in __call__
    _run_app(opts, args, bad_lines)
  File "/usr/bin/../lib/kitty/kitty/main.py", line 156, in _run_app
    window_id = create_os_window(
  File "/usr/bin/../lib/kitty/kitty/main.py", line 91, in load_all_shaders
    def load_all_shaders(semi_transparent: bool = False) -> None:
KeyboardInterrupt
```

Therefore I have some ideas after benchmarking some runs:

* Consider caching the shader programs. I believe this will help the most.
* Consider caching a reasonably large sprite map for the current font on startup.
* Consider caching parts of the fonts themselves since we have a very simple rendering pipeline. Being something of a font format expert I'm happy to help if this ends up being needed.
* Consider compiling Kitty's Python scripts with [Nuitka](https://nuitka.net/). I've used Nuitka with other projects and have gotten considerable speed gains out of it in the past. This eliminates the step of starting an interpreter.
* Consider, if the above idea fails, Cython.
* Consider allowing kitty windows to have different font sizes, thus eliminating one of the main reasons you'd even want to run simultaneous concurrent Kitty instances. (This is the only reason I do, I welcome @primalmotion to chime in with other reasons separate instances are needed.)

I will do more testing and hope this will lead to good discussion and patches so we can close the performance gap.

**Comment by kovidgoyal** [comment]
I really doubt shader compilation is the bottleneck. How are you
measuring this? And there is no good reason to not use multiple windows
in a single process (individual OS windows can have their own font
sizes, that ability was added years ago). Though am sure your Alacritty
apologist will come back with, but if it crashes all the windows will
go. That's the standard refrain people that use it as an excuse use.

As far actual startup time difference is concerned, as far as I remember
from the last time I looked at it, the difference in startup time
between kitty and alacritty is on the order of 0.1 seconds. Anybody that
is sensitive to that should really not be using alacritty at all. xterm
for instance starts up an order of magnitude faster than alacritty. So
the next time you get trolled about how alacritty starts faster than
kitty, suggest the OP switches to xterm.

Basically here is the startup time skinny:

1) Initializing OpenGL takes on the order of 0.1 seconds. This is common
to all OpenGL based terminal emulators and makes then all an order of
magnitude slower than xterm to start.

2) The remaining 0.1 second of difference between kitty and alacritty
comes from loading python code at startup.

3) kitty has --single-instance which means it starts up faster than
anything else that does not have its own client/server mode. 

I suggest you dont worry about it. In my experience, people that trot
out this canard are just trying to justify their terminal emulator
choice which for some reason is closely bound up with their identity.

That said, I am happy to review PRs to improve startup performance.

**Comment by primalmotion** [comment]
Happy that led to some investigations. I obviously gave it another go a few hours ago (just to check if I was not actually telling bs :)), and I confirm that for me, it is often slower, but sometimes equivalent.

About having separate processes, it's mainly because I can easily sandbox everything I run inside without impacting anything else.

**Comment by kovidgoyal** [comment]
Closing, but feel free to continue the discussion.

**Comment by primalmotion** [comment]
https://github.com/kovidgoyal/kitty/issues/4292#issuecomment-980857020

And that is also, while far from the main one, at least one of the reasons I don't use kitty. I really don't like how you appear online... you're flying thousands miles above the ground

--
His Alacritty apologist

**Comment by kovidgoyal** [comment]
@primalmotion Live long and prosper

**Comment by ctrlcctrlv** [comment]
No, sometimes kitty takes much longer to start. Compiling the shaders sometimes takes quite a while. I've gotten output from `time` of even up to 1.5s, 2s to start a new process.

Also, how can different kitty windows have different sizes, for my information @kovidgoyal ? I don't really much care about pleasing this user, I just want to fix a bug I found.

**Comment by kovidgoyal** [comment]
On Sat, Nov 27, 2021 at 11:33:32PM -0800, Fredrick Brennan wrote:
> No, sometimes kitty takes much longer to start. Compiling the shaders sometimes takes quite a while. I've gotten output from `time` of even up to 1.5s, 2s to start a new process.

That would be most interesting indeed. I have never seen that on my
system. It's likely the cache used for shaders on your system is getting
full causing a recompile. The cache is managed by the OpenGL
drivers/mesa not kitty. Are you running lots of shader heavy programs?

> 
> Also, how can different kitty windows have different sizes, for my information @kovidgoyal ? I don't really much care about pleasing this user, I just want to fix a bug I found.


https://sw.kovidgoyal.net/kitty/conf/#font-sizes

**Comment by kovidgoyal** [comment]
And if you want to control the font size via remote control: https://sw.kovidgoyal.net/kitty/remote-control/#kitty-set-font-size

**Comment by kovidgoyal** [comment]
On my system with
```
export MESA_GLSL_CACHE_DISABLE=true
time kitty false                                                                                                        

real	0.304
user	0.242
sys	0.055
maxmem	91 MB
faults	4
```

```
unset MESA_GLSL_CACHE_DISABLE
time kitty false                                                                                                                         

real	0.247
user	0.177
sys	0.061
maxmem	90 MB
faults	4
```

**Comment by ctrlcctrlv** [comment]
Hey @kovidgoyal, I did some more testing to try to figure out what's causing the 2 seconds. I agree, it's not shaders. The reason I thought that is because I misread some output from `perf`, wrongly thinking I was tracking clock time but had forgotten `-e time-clock`. Window creation _is_ the problem, though.

![flame](https://user-images.githubusercontent.com/838783/143734966-466c5129-366f-43b6-a1f9-65caac7b4be5.png)

The big bar is caused by this call, before we start the `Boss`:

https://github.com/kovidgoyal/kitty/blob/346ea8c8a026727dec6a7f60a3f07e592c11e265/kitty/main.py#L162-L166

**Comment by kovidgoyal** [comment]
Yeah that will in turn be caused by some underlying system call being slow. You can trace that further in glfw.c where create_os_window is defined. Most likely it will be glfwCreateWindow which you can trace some more in glfw/x11_*.c

**Comment by ctrlcctrlv** [comment]
This is under Wayland, by the way :) I'll add some debug printf's.

**Comment by ctrlcctrlv** [comment]
(Also I'm sorry for behavior of @primalmotion, your repository rules regarding which issues stay open (closed until full details regarding reproduction) are fine by me, and you even welcomed discussion. I think this doesn't affect many people (indeed, I'd never noticed because I use single instance mode) but I do want to put in the time to fix it if possible because even equivalent time to alacritty means we're missing out on some possible speed increase, I want to maintain the project's record in all metrics :))

**Comment by ctrlcctrlv** [comment]
All right, here's some more information @kovidgoyal 
```
[332 03:45:49.639477] Setting window hints…
[332 03:45:49.639496] Done setting window hints…
[332 03:45:49.639498] get_window_content_scale
[332 03:45:49.639501] load_fonts_data
[332 03:45:49.642291] done load_fonts_data
[332 03:45:49.642308] Running glfwWindowHint
[332 03:45:49.642310] Running glfwCreateWindow
[332 03:45:50.990351] Would destroy window here on X11
[332 03:45:50.990380] glfwMakeContextCurrent
[332 03:45:50.990636] glfwMakeContextCurrent done
[332 03:45:50.990659] gl_init
[332 03:45:50.990969] gl_init done
[332 03:45:50.990975] blank_canvas
[332 03:45:50.990983] blank_canvas done
[332 03:45:50.990994] glfwSwapBuffers
[332 03:45:50.997449] glfwSwapBuffers done
[332 03:45:50.997515] glfwSetInputMode
[332 03:45:50.997521] glfwSetInputMode done
[332 03:45:50.997525] Loading shader programs…
[332 03:45:51.012992] Done loading shader programs…
```
Note the large gap when we call `glfwCreateWindow`. That's the problem somehow. Ideas?

```diff
diff --git a/kitty/glfw.c b/kitty/glfw.c
index 84b545d5..aeb89894 100644
--- a/kitty/glfw.c
+++ b/kitty/glfw.c
@@ -708,6 +708,7 @@ create_os_window(PyObject UNUSED *self, PyObject *args, PyObject *kw) {
 
     static bool is_first_window = true;
     if (is_first_window) {
+	log_error("Setting window hints…");
         glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, OPENGL_REQUIRED_VERSION_MAJOR);
         glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, OPENGL_REQUIRED_VERSION_MINOR);
         glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
@@ -741,6 +742,7 @@ create_os_window(PyObject UNUSED *self, PyObject *args, PyObject *kw) {
     // creation, which causes a resize event and all the associated processing.
     // The temp window is used to get the DPI.
     glfwWindowHint(GLFW_VISIBLE, false);
+    log_error("Done setting window hints…");
     GLFWwindow *common_context = global_state.num_os_windows ? global_state.os_windows[0].handle : NULL;
     GLFWwindow *temp_window = NULL;
 #ifdef __APPLE__
@@ -757,10 +759,13 @@ create_os_window(PyObject UNUSED *self, PyObject *args, PyObject *kw) {
     }
     float xscale, yscale;
     double xdpi, ydpi;
+    log_error("get_window_content_scale");
     get_window_content_scale(temp_window, &xscale, &yscale, &xdpi, &ydpi);
+    log_error("load_fonts_data");
     FONTS_DATA_HANDLE fonts_data = load_fonts_data(OPT(font_size), xdpi, ydpi);
     PyObject *ret = PyObject_CallFunction(get_window_size, "IIddff", fonts_data->cell_width, fonts_data->cell_height, fonts_data->logical_dpi_x, fonts_data->logical_dpi_y, xscale, yscale);
     if (ret == NULL) return NULL;
+    log_error("done load_fonts_data");
     int width = PyLong_AsLong(PyTuple_GET_ITEM(ret, 0)), height = PyLong_AsLong(PyTuple_GET_ITEM(ret, 1));
     Py_CLEAR(ret);
     // The GLFW Wayland backend cannot create and show windows separately so we
@@ -768,23 +773,36 @@ create_os_window(PyObject UNUSED *self, PyObject *args, PyObject *kw) {
     // https://github.com/glfw/glfw/issues/1268 It doesn't matter since there
     // is no startup notification in Wayland anyway. It amazes me that anyone
     // uses Wayland as anything other than a butt for jokes.
+    log_error("Running glfwWindowHint");
     if (global_state.is_wayland) glfwWindowHint(GLFW_VISIBLE, true);
+    log_error("Running glfwCreateWindow");
     GLFWwindow *glfw_window = glfwCreateWindow(width, height, title, NULL, temp_window ? temp_window : common_context);
+    log_error("Would destroy window here on X11");
     if (temp_window) { glfwDestroyWindow(temp_window); temp_window = NULL; }
     if (glfw_window == NULL) { PyErr_SetString(PyExc_ValueError, "Failed to create GLFWwindow"); return NULL; }
+    log_error("glfwMakeContextCurrent");
     glfwMakeContextCurrent(glfw_window);
+    log_error("glfwMakeContextCurrent done");
     if (is_first_window) {
+        log_error("gl_init");
         gl_init();
+        log_error("gl_init done");
     }
     bool is_semi_transparent = glfwGetWindowAttrib(glfw_window, GLFW_TRANSPARENT_FRAMEBUFFER);
     // blank the window once so that there is no initial flash of color
     // changing, in case the background color is not black
+    log_error("blank_canvas");
     blank_canvas(is_semi_transparent ? OPT(background_opacity) : 1.0f, OPT(background));
+    log_error("blank_canvas done");
 #ifndef __APPLE__
     if (is_first_window) glfwSwapInterval(OPT(sync_to_monitor) && !global_state.is_wayland ? 1 : 0);
 #endif
+    log_error("glfwSwapBuffers");
     glfwSwapBuffers(glfw_window);
+    log_error("glfwSwapBuffers done");
+    log_error("glfwSetInputMode");
     glfwSetInputMode(glfw_window, GLFW_LOCK_KEY_MODS, true);
+    log_error("glfwSetInputMode done");
     if (!global_state.is_wayland) {
         PyObject *pret = PyObject_CallFunction(pre_show_callback, "N", native_window_handle(glfw_window));
         if (pret == NULL) return NULL;
@@ -803,7 +821,9 @@ create_os_window(PyObject UNUSED *self, PyObject *args, PyObject *kw) {
 #endif
     }
     if (is_first_window) {
+	log_error("Loading shader programs…");
         PyObject *ret = PyObject_CallFunction(load_programs, "O", is_semi_transparent ? Py_True : Py_False);
+	log_error("Done loading shader programs…");
         if (ret == NULL) return NULL;
         Py_DECREF(ret);
 #define CC(dest, shape) {\
```

**Comment by ctrlcctrlv** [comment]
Hello @kovidgoyal 

I have determined that the function `_glfwInitEGL` takes over a second on my system. Compare the lines containing it in this output:

I can post the long-ish patch but it's as boring as you'd expect.

```
[fred@🍇葡萄🍇kitty-master-2021]$ vim glfw/wl_window.c [fred@🍇葡萄🍇kitty-master-2021]$ OVERRIDE_CFLAGS="`pkg-config --shared --libs --cflags python3` $PWD/kitty/fast_data_types.so" make debug
python3 setup.py build  --debug
CC: ['gcc'] (11, 1)
[1/1] Compiling [wayland] glfw/wl_window.c ... done
[1/1] Linking [wayland] kitty/glfw-wayland ... done
[fred@🍇葡萄🍇kitty-master-2021]$ HOME=`mktemp -d` austin -i 10 --pipe ./kitty/launcher/kitty --instance-group `date +%s$RANDOM` echo > austin.txt
[332 04:33:55.237263] Setting window hints…
[332 04:33:55.237284] Done setting window hints…
[332 04:33:55.237286] get_window_content_scale
[332 04:33:55.237288] load_fonts_data
[332 04:33:55.239817] done load_fonts_data
[332 04:33:55.239834] Running glfwWindowHint
[332 04:33:55.239837] Running glfwCreateWindow
[332 04:33:55.239845] initialize_csd_metrics(window);
[332 04:33:55.239848] initialize_csd_metrics(window); done
[332 04:33:55.239849] window->wl.transparent = fbconfig->transparent;
[332 04:33:55.239850] window->wl.transparent = fbconfig->transparent; done
[332 04:33:55.239852] strncpy(window->wl.appId, wndconfig->wl.appId, sizeof(window->wl.appId));
[332 04:33:55.239853] strncpy(window->wl.appId, wndconfig->wl.appId, sizeof(window->wl.appId)); done
[332 04:33:55.239855] if (!createSurface(window, wndconfig))
[332 04:33:55.239866]     return false; done
[332 04:33:55.239869] if (ctxconfig->client != GLFW_NO_API)
[332 04:33:55.239870] if (ctxconfig->client != GLFW_NO_API) done
[332 04:33:55.239872]     if (ctxconfig->source == GLFW_EGL_CONTEXT_API ||
[332 04:33:55.239874]         ctxconfig->source == GLFW_NATIVE_CONTEXT_API)
[332 04:33:55.239875]     if (ctxconfig->source == GLFW_EGL_CONTEXT_API || done
[332 04:33:55.239877]         ctxconfig->source == GLFW_NATIVE_CONTEXT_API) done
[332 04:33:55.239878]         if (!_glfwInitEGL())
[332 04:33:56.571228]             return false; done
[332 04:33:56.571248]         if (!_glfwCreateContextEGL(window, ctxconfig, fbconfig))
[332 04:33:56.572743]             return false; done
[332 04:33:56.572787]     else if (ctxconfig->source == GLFW_OSMESA_CONTEXT_API)
[332 04:33:56.572790]     } done
[332 04:33:56.572791] }
[332 04:33:56.572792] } done
[332 04:33:56.572798] if (wndconfig->title)
[332 04:33:56.572799] if (wndconfig->title) done
[332 04:33:56.572801]     window->wl.title = _glfw_strdup(wndconfig->title);
[332 04:33:56.572805]     window->wl.title = _glfw_strdup(wndconfig->title); done
[332 04:33:56.572807] if (wndconfig->maximized)
[332 04:33:56.572809]     window->wl.maximize_on_first_show = true; done
[332 04:33:56.572810] if (wndconfig->visible)
[332 04:33:56.572813] if (wndconfig->visible) done
[332 04:33:56.572815]     if (!createXdgSurface(window))
[332 04:33:56.590741]         return false; done
[332 04:33:56.590762]     window->wl.visible = true;
[332 04:33:56.590764]     window->wl.visible = true; done
[332 04:33:56.590767] } else {
[332 04:33:56.590769] } done
[332 04:33:56.590777] window->wl.currentCursor = NULL;
[332 04:33:56.590779] window->wl.currentCursor = NULL; done
[332 04:33:56.590781] // Don't set window->wl.cursorTheme to NULL here.
[332 04:33:56.590785] // Don't set window->wl.cursorTheme to NULL here. done
[332 04:33:56.590787] window->wl.monitors = calloc(1, sizeof(_GLFWmonitor*));
[332 04:33:56.590790] window->wl.monitors = calloc(1, sizeof(_GLFWmonitor*)); done
[332 04:33:56.590792] window->wl.monitorsCount = 0;
[332 04:33:56.590794] window->wl.monitorsCount = 0; done
[332 04:33:56.590795] window->wl.monitorsSize = 1;
[332 04:33:56.590800] window->wl.monitorsSize = 1; done
[332 04:33:56.590802] return true;
[332 04:33:56.597699] Would destroy window here on X11
[332 04:33:56.597721] glfwMakeContextCurrent
[332 04:33:56.598051] glfwMakeContextCurrent done
[332 04:33:56.598070] gl_init
[332 04:33:56.598386] gl_init done
[332 04:33:56.598398] blank_canvas
[332 04:33:56.598407] blank_canvas done
[332 04:33:56.598415] glfwSwapBuffers
[332 04:33:56.604779] glfwSwapBuffers done
[332 04:33:56.604807] glfwSetInputMode
[332 04:33:56.604812] glfwSetInputMode done
[332 04:33:56.604815] Loading shader programs…
[332 04:33:56.623007] Done loading shader programs…
```

**Comment by kovidgoyal** [comment]
that's in egl_context.c, you can debug it further there.

**Comment by kovidgoyal** [comment]
> it if possible because even equivalent time to alacritty means we're missing out on some possible speed increase, I want to maintain the project's record in all metrics :))

by all means go for it, I will help as I can,

**Comment by ctrlcctrlv** [comment]
Narrowed it down some more. `eglGetPlatformDisplayEXT` is taking the time. That's just `GetProcAddress`??

```
[332 06:39:47.859392] Searching for EGL_EXT_platform_base
[332 06:39:47.859397] Searching for EGL_EXT_platform_x11
[332 06:39:47.859400] Searching for EGL_EXT_platform_wayland
[332 06:39:47.859402] Searching for EGL_ANGLE_platform_angle
[332 06:39:47.859404] Searching for EGL_ANGLE_platform_angle_opengl
[332 06:39:47.859405] Searching for EGL_ANGLE_platform_angle_d3d
[332 06:39:47.859407] Searching for EGL_ANGLE_platform_angle_vulkan
[332 06:39:47.859408] Searching for EGL_ANGLE_platform_angle_metal
[332 06:39:47.859410] eglGetProcAddress("eglGetPlatformDisplayEXT");
[332 06:39:47.859413] eglGetProcAddress("eglGetPlatformDisplayEXT"); OK
[332 06:39:47.859415] eglGetProcAddress("eglCreatePlatformWindowSurfaceEXT");
[332 06:39:47.859417] eglGetProcAddress("eglCreatePlatformWindowSurfaceEXT"); OK
[332 06:39:47.859421] eglGetPlatformDisplayEXT 12760
[332 06:39:49.136985] eglGetPlatformDisplayEXT 12760 OK
[332 06:39:49.137009] free(attribs)
[332 06:39:49.137011] free ok
```

**Comment by ctrlcctrlv** [comment]
When I do an `strace` I find that the one second or so pause happens here:

```
openat(AT_FDCWD, "/dev/nvidiactl", O_RDWR) = 6
fcntl(6, F_SETFD, FD_CLOEXEC)           = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xd2, 0x48), 0x7ffccedd01b0) = 0
openat(AT_FDCWD, "/sys/devices/system/memory/block_size_bytes", O_RDONLY) = 7
read(7, "8000000\n", 99)                = 8
close(7)                                = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xd6, 0x8), 0x7ffccedd0240) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xc8, 0x900), 0x7f24b70aba00) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2b, 0x20)^C^C^C^Cstrace: Process 1269523 detached
 <detached ...>
```

It's such a long pause I can ^C^C^C.

Running as:

```bash
HOME=`mktemp -d` strace ./kitty/launcher/kitty --instance-group debug`dd if=/dev/random count=6 bs=1 status=none|base64` echo
```

**Comment by ctrlcctrlv** [comment]
This kind of fills me with dread because from the context of the ioctl it looks like we're inside NVIDIA internal driver code:

```c++
openat(AT_FDCWD, "/proc/driver/nvidia/params", O_RDONLY) = 6
newfstatat(6, "", {st_mode=S_IFREG|0444, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(6, "ResmanDebugLevel: 4294967295\nRmL"..., 1024) = 804
close(6)                                = 0
stat("/dev/nvidiactl", {st_mode=S_IFCHR|0666, st_rdev=makedev(0xc3, 0xff), ...}) = 0
openat(AT_FDCWD, "/dev/nvidiactl", O_RDWR) = 6
fcntl(6, F_SETFD, FD_CLOEXEC)           = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xd2, 0x48), 0x7ffce5335230) = 0
openat(AT_FDCWD, "/sys/devices/system/memory/block_size_bytes", O_RDONLY) = 7
read(7, "8000000\n", 99)                = 8
close(7)                                = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xd6, 0x8), 0x7ffce53352c0) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xc8, 0x900), 0x7f7049015a00) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2b, 0x20), 0x7ffce5335390) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffce5335360) = 0
openat(AT_FDCWD, "/tmp/tmp.iD2gNca7ny/.nv/nvidia-application-profile-globals-rc", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/tmp.iD2gNca7ny/.nv/nvidia-application-profiles-rc", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/tmp.iD2gNca7ny/.nv/nvidia-application-profiles-rc.d", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/nvidia/nvidia-application-profiles-rc", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/nvidia/nvidia-application-profiles-rc.d/", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/nvidia/nvidia-application-profiles-495.44-rc", O_RDONLY) = 7
fstat(7, {st_mode=S_IFREG|0644, st_size=9279, ...}) = 0
brk(0x564d3d359000)                     = 0x564d3d359000
newfstatat(7, "", {st_mode=S_IFREG|0644, st_size=9279, ...}, AT_EMPTY_PATH) = 0
read(7, "# Application profiles for the N"..., 8192) = 8192
read(7, "    { \"pattern\": \"RocketLeague\","..., 4096) = 1087
close(7)                                = 0
openat(AT_FDCWD, "/usr/share/nvidia/nvidia-application-profiles-rc", O_RDONLY) = -1 ENOENT (No such file or directory)
getpid()                                = 1269449
readlink("/proc/1269449/exe", "/home/fred/Workspace/kitty-maste"..., 4095) = 59
```

**Comment by ctrlcctrlv** [comment]
Maybe a known NVIDIA bug: https://forums.developer.nvidia.com/t/very-long-latency-2-6s-2-6s-when-running-cuda-opencl-code-on-threadripper-linux-server/147119

**Comment by ctrlcctrlv** [comment]
Everything I'm reading is implicating CUDA and OpenCL. We don't use either, can't we just run without?

**Comment by ctrlcctrlv** [comment]
Oh, good news :-D

```
$ sudo systemctl enable --now nvidia-persistenced
```

This fixed it for me. Can I add a FAQ item @kovidgoyal?

**Comment by ctrlcctrlv** [comment]
Or at least, the lag on startup is much more infrequent with the nvidia-persistenced running, but not gone :frowning: 

I wonder if there's something I can call to say "we have no OpenCL/CUDA"

**Comment by ctrlcctrlv** [comment]
Okay, I figured it out for real this time. This explanation explains everything.

```
# echo on > /sys/bus/pci/devices/0000\:01\:00.0/power/control
```

Fixes it. This disables software-level power management of NVIDIA card. NVIDIA card will go to sleep if not being used. In multi GPU system, this will happen a lot. But if you open kitty, you have to wake up GPU.

Also persistenced does help I'm sure but this is the real solution. It explains why I kept thinking I needed random $HOME dir or random instance group but neither made it 100% reproducable. My card was turning off after a short delay, that's why.

Note that your card may be at different location. Do `lspci` to find it. My output:

```
[fred@🍇葡萄🍇~]$ sudo lspci | rg NVIDIA
01:00.0 VGA compatible controller: NVIDIA Corporation GA106M [GeForce RTX 3060 Mobile / Max-Q] (rev a1)
01:00.1 Audio device: NVIDIA Corporation Device 228e (rev a1)
```

You want the VGA one obviously. FIXED

**Comment by kovidgoyal** [comment]
Nice, that was some productive debugging. Not sure we can do anything about it though.

**Comment by kovidgoyal** [comment]
Though I do wonder why only kitty (well technically glfw) triggers this behavior.

**Comment by ctrlcctrlv** [comment]
I also wonder, but remember that in the beginning of the issue I tested Alacritty thruogh XWayland because I couldn't get it working at all in `kwin_wayland`. Perhaps XWayland knows not to wake up NVIDIA GPU that's not in use.

**Comment by ctrlcctrlv** [comment]
Although, Alacritty is not making the offending `ioctl`. But then again I don't think Alacritty uses GLSL.

```
openat(AT_FDCWD, "/proc/driver/nvidia/params", O_RDONLY) = 14
newfstatat(14, "", {st_mode=S_IFREG|0444, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(14, "ResmanDebugLevel: 4294967295\nRmL"..., 1024) = 804
close(14)                               = 0
stat("/dev/nvidiactl", {st_mode=S_IFCHR|0666, st_rdev=makedev(0xc3, 0xff), ...}) = 0
openat(AT_FDCWD, "/dev/nvidiactl", O_RDWR) = 14
fcntl(14, F_SETFD, FD_CLOEXEC)          = 0
ioctl(14, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xd2, 0x48), 0x7ffd0b4e78c0) = 0
openat(AT_FDCWD, "/sys/devices/system/memory/block_size_bytes", O_RDONLY) = 16
read(16, "8000000\n", 99)               = 8
close(16)                               = 0
ioctl(14, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xd6, 0x8), 0x7ffd0b4e7950) = 0
ioctl(14, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0xc8, 0x900), 0x7f13b0bb7a00) = 0
ioctl(14, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2b, 0x20), 0x7ffd0b4e7a20) = 0
ioctl(14, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffd0b4e79f0) = 0
```

**Comment by ctrlcctrlv** [comment]
No, wait a minute. Alacritty is also affected, I just confirmed it. @kovidgoyal the reason it _SEEMED_ not to be affected is because I ran it right after kitty, so it benefited from kitty having done the hard work of waking up the GPU (it's like a bear, quick to hibernate, slow to wake, but once awake, ready to kill you)

**Comment by kovidgoyal** [comment]
Yes, I was very surprised that this affected only kitty. Creating an EGL surface is a very generic operation.

**Comment by ctrlcctrlv** [comment]
I think that because most videogames (and Blender) already take so long to start up, this problem is really only felt by users of GPU terminal emulators, which are the rare GPU program one expects to rapidly start. I noticed that most people complaining about it were talking about CUDA or OpenCL, that was a red herring, it's just that _in those applications_ slow start is more noticeable, nothing at all to do with whether CUDA actually in use. :)

**Comment by ctrlcctrlv** [comment]
Oh, before I leave the issue for now, here is a new flamegraph:

![image](https://user-images.githubusercontent.com/838783/143770284-c3867c84-b660-4db7-9090-bd160d9eab4f.png)

So, if we want further cuts in launch time, should such be desirable as another metric to be the best in (I think all metrics are :wink:), then the two lowest hanging fruit remain interpreter startup and GLSL shader program caching.

@kovidgoyal Have you considered Cython or Nuitka at all?

**Comment by ctrlcctrlv** [comment]
Apparently 6% of startup time is spent on shader compilation:

![image](https://user-images.githubusercontent.com/838783/143770454-bbe8d611-070a-4c87-9cc4-76d411d4f59a.png)

**Comment by kovidgoyal** [comment]
On Sun, Nov 28, 2021 at 05:44:38AM -0800, Fredrick Brennan wrote:
> Apparently 6% of startup time is spent on shader compilation:
> 
> ![image](https://user-images.githubusercontent.com/838783/143770454-bbe8d611-070a-4c87-9cc4-76d411d4f59a.png)

Shader compilation is already cached automatically by mesa. is that not
functional on your system?

**Comment by kovidgoyal** [comment]
On Sun, Nov 28, 2021 at 05:41:11AM -0800, Fredrick Brennan wrote:
> Oh, before I leave the issue for now, here is a new flamegraph:
> 
> ![image](https://user-images.githubusercontent.com/838783/143770284-c3867c84-b660-4db7-9090-bd160d9eab4f.png)
> 
> So, if we want further cuts in launch time, should such be desirable as another metric to be the best in (I think all metrics are :wink:), then the two lowest hanging fruit remain interpreter startup and GLSL shader program caching.
> 
> @kovidgoyal Have you considered Cython or Nuitka at all?

No, kitty's startup time is good enough for me. But if you wish to
investigate, feel free. IMO '# of awesome things it does that no one else
does' is a more important metric.

**Comment by mindrunner** [comment]
Sorry for jumping back in here. I just migrated from alacrity to kitty and figured out, that kitty is waking up my nvidia card each time i fire up a terminal resulting in a serious lag. I really wonder why kitty wants to wake up the nvidia card, even though the whole sway session is running on intel and it is not started with `prime-run` or similar. This looks like a bug and seeing the ticket being closed is kinda sad. 

I like the tab-features and stuff, but the startup delay renders kitty absolutely unusable for me. :( Really nothing you can do about it?

**Comment by ctrlcctrlv** [comment]
@mindrunner A work-around was added to Kitty's documentation in #4296: https://github.com/kovidgoyal/kitty/blob/master/docs/faq.rst#why-does-kitty-sometimes-start-slowly-on-my-linux-system

It'll likely appear on the Kitty website next release.

**Comment by mindrunner** [comment]
Saved my life! Thanks!

Most people use intel/nvidia combinations, though. Maybe it is a good idea to give them the right parameters in documentation as well. For me it is:

```
MESA_LOADER_DRIVER_OVERRIDE=i965 __EGL_VENDOR_LIBRARY_FILENAMES=/usr/share/glvnd/egl_vendor.d/50_mesa.json kitty
```

**Comment by ctrlcctrlv** [comment]
I don't think that's the best idea, as [different Intel chipsets may have different drivers](https://wiki.archlinux.org/title/OpenGL#Installation), and there's also the Mesa `iris` driver. It's kind of a good thing my setup is uncommon, as it requires people to _read the text_ and do the legwork you did to determine `i965`. My `radeonsi` is just an example of something that _can_ go there, not expected to necessarily be what _you_ need to put there.

**Comment by ctrlcctrlv** [comment]
By the way, you're definitely right that AMD/NVIDIA is a very unusual combination, especially in a laptop. You'd usually expect Radeon graphics on CPU and a more powerful AMD Radeon GPU, so radeon/radeon. All the bugs I've faced/found is enough to tell me that!. ;-)

**Comment by arsenerei** [comment]
@ctrlcctrlv I know this is late, but do you mind sharing your `perf` command for this comparison? I tried `perf record kitty` and `perf record --call-graph=dwarf kitty`, but running `perf script | ./stackcollapse.pl | ./flamegraph.pl > perf.svg` outputs `ERROR:` No stack counts found`.

**Comment by ctrlcctrlv** [comment]
Sorry, I do mind, it took a while to figure out and I didn't write it down, and I don't feel like doing it again. I can say that you're on the wrong track using regular GNU perf though, I tried that too. I used https://github.com/P403n1x87/austin and I think cargo-flamegraph. Exact command is an exercise to the reader :stuck_out_tongue_closed_eyes:
